### PR TITLE
`src/cdef_apply.rs`: Deduplicate w/ generics

### PR DESCRIPTION
**Pull request by thedataking**

### CODE REVIEW

File: src/cdef_apply.rs [hunk 1]
```diff
@@ -10,105 +11,123 @@ use crate::src::internal::Rav1dDSPContext;
 use crate::src::internal::Rav1dFrameContext;
 use crate::src::internal::Rav1dTaskContext;
 use crate::src::lf_mask::Av1Filter;
+
 use libc::memcpy;
```
Comment (kkysen) [inline src/cdef_apply.rs:15]:
```suggestion
use libc::memcpy;
```

No import newlines.

File: src/cdef_apply.rs [hunk 2]
```diff
@@ -10,105 +11,123 @@ use crate::src::internal::Rav1dDSPContext;
...
+            (-2 * y_stride) as usize
+        } else {
+            (-2 * y_stride << 1) as usize
+        };
         memcpy(
             (*dst.offset(0)).offset(y_stride as isize) as *mut c_void,
             (*src.offset(0)).offset((7 as c_int as isize * y_stride) as isize) as *const c_void,
-            (-(2 as c_int) as isize * y_stride << 1) as usize,
+            n,
         );
```
Comment (kkysen) [inline src/cdef_apply.rs:44]: Looks like a `pixel_copy`.

**Review by kkysen — CHANGES_REQUESTED** [review]
Same comments as #586 and #867, plus we can make the call to it in `recon.rs` generic, too (if I missed that in the other PRs, this comment would apply there as well).

**Comment by thedataking** [comment]
I found a couple more missing pixel copies and tested the result against the full argon suite locally. Advance apologies if I missed something else you pointed out in previous reviews.

File: src/cdef_apply.rs [hunk 3]
```diff
@@ -536,9 +568,11 @@ pub(crate) unsafe fn rav1d_cdef_brow_16bpc(
...
             );
         }
-        ptrs[0] = (ptrs[0]).offset((8 * PXSTRIDE((*f).cur.stride[0])) as isize);
-        ptrs[1] = (ptrs[1]).offset((8 * PXSTRIDE((*f).cur.stride[1]) >> ss_ver) as isize);
-        ptrs[2] = (ptrs[2]).offset((8 * PXSTRIDE((*f).cur.stride[1]) >> ss_ver) as isize);
+        ptrs[0] = (ptrs[0]).offset((8 * BD::pxstride((*f).cur.stride[0] as usize)) as isize);
+        ptrs[1] =
+            (ptrs[1]).offset((8 * BD::pxstride((*f).cur.stride[1] as usize) >> ss_ver) as isize);
+        ptrs[2] =
+            (ptrs[2]).offset((8 * BD::pxstride((*f).cur.stride[1] as usize) >> ss_ver) as isize);
```
Comment (kkysen) [inline src/cdef_apply.rs:575]:
```suggestion
        ptrs[0] = (ptrs[0]).offset(8 * BD::pxstride((*f).cur.stride[0] as usize) as isize);
        ptrs[1] =
            (ptrs[1]).offset(8 * BD::pxstride((*f).cur.stride[1] as usize) as isize >> ss_ver);
        ptrs[2] =
            (ptrs[2]).offset(8 * BD::pxstride((*f).cur.stride[1] as usize) as isize >> ss_ver);
```

I've probably messed this up elsewhere myself, but signed/arithmetic and unsigned/logical shifts are different.

File: src/recon.rs [hunk 4]
```diff
@@ -4571,35 +4565,19 @@ pub(crate) unsafe fn rav1d_filter_sbrow_cdef<BD: BitDepth>(tc: &mut Rav1dTaskCon
...
-                prev_mask,
-                start - 2,
-                start,
-                1 as c_int,
-                sby,
-            ),
-        };
+        rav1d_cdef_brow::<BD>(
+            tc,
+            p_up.as_mut_ptr().cast(),
```
Comment (kkysen) [inline src/recon.rs:4570]:
```suggestion
            p_up.as_mut_ptr(),
```

File: src/recon.rs [hunk 5]
```diff
@@ -4571,35 +4565,19 @@ pub(crate) unsafe fn rav1d_filter_sbrow_cdef<BD: BitDepth>(tc: &mut Rav1dTaskCon
...
     }
     let n_blks = sbsz - 2 * ((sby + 1) < (*f).sbh) as c_int;
     let end = cmp::min(start + n_blks, (*f).bh);
-    match BD::BPC {
-        BPC::BPC8 => rav1d_cdef_brow_8bpc(tc, p.as_ptr().cast(), mask, start, end, 0 as c_int, sby),
-        BPC::BPC16 => {
-            rav1d_cdef_brow_16bpc(tc, p.as_ptr().cast(), mask, start, end, 0 as c_int, sby)
-        }
-    };
+    rav1d_cdef_brow::<BD>(tc, p.as_ptr().cast(), mask, start, end, 0 as c_int, sby);
```
Comment (kkysen) [inline src/recon.rs:4580]:
```suggestion
    rav1d_cdef_brow::<BD>(tc, p.as_ptr(), mask, start, end, 0 as c_int, sby);
```

**Review by kkysen — CHANGES_REQUESTED** [review]

File: src/cdef_apply.rs [hunk 6]
```diff
@@ -101,14 +122,17 @@ unsafe fn backup2x8(
     if flag as c_uint & BACKUP_2X8_Y as c_int as c_uint != 0 {
         let mut y = 0;
         while y < 8 {
-            memcpy(
-                ((*dst.offset(0))[y as usize]).as_mut_ptr() as *mut c_void,
-                &mut *(*src.offset(0)).offset((y_off + x_off as isize - 2 as isize) as isize)
-                    as *mut pixel as *const c_void,
-                ((2 as c_int) << 1) as usize,
+            BD::pixel_copy(
+                slice::from_raw_parts_mut(((*dst.offset(0))[y as usize]).as_mut_ptr(), 2),
```
[the remark this answers is not in the document]
Comment (thedataking) [reply src/cdef_apply.rs:126]: right, good catch

File: src/cdef_apply.rs (hunk 3 above)
Comment (thedataking) [reply src/cdef_apply.rs:575]: indeed, good catch!

**Review by kkysen — APPROVED** [review]

**Pull request merged**